### PR TITLE
Rebased docker/distribution

### LINKED ISSUE #4339: Update docker registry package
https://github.com/docker/distribution/releases

When that time comes, [code.google.com/p/go-uuid/uuid](https://github.com/openshift/origin/tree/master/Godeps/_workspace/src/code.google.com/p/go-uuid/uuid) which is a duplicate of [https://github.com/pborman/uuid](https://github.com/openshift/origin/tree/master/Godeps/_workspace/src/github.com/pborman/uuid) can be removed from our deps.

### LINKED ISSUE #4672: OpenStack Swift support for docker-registry
To get OpenStack Swift support for the docker-registry we would need to update the deps to 2.1.1 since this was merged last July :

https://github.com/docker/distribution/pull/493

support for OpenStack Swift would allow for nice things like HA for on premise clouds

### PR DESCRIPTION
**Pull request by miminar**
Resolves #4339
Resolves #4672
Partially addresses #5900 (this PR just reenables liveness probe).

Benefits:
- [Storage support openstack swift](https://github.com/docker/distribution/pull/493)
- less painful consequent rebase

Notable features by Docker:
- [Support for pull-through cached mirroring](https://github.com/docker/distribution/issues/19)
- [Implemented soft deletes of manifests and blobs through the Registry HTTP API](https://github.com/docker/distribution/issues/461)
- Added repository catalog API to Registry HTTP JSON API
- Read-only mode (required config change and restart)
- Complete list new features:
  -  [v2.1](https://github.com/docker/distribution/issues?utf8=%E2%9C%93&q=milestone%3ARegistry%2F2.1+label%3Afeature+)
  -  [v2.2](https://github.com/docker/distribution/issues?utf8=%E2%9C%93&q=milestone%3ARegistry%2F2.2+label%3Afeature)

### CODE REVIEW

**Comment by miminar** [comment]
Ping @kargakis, @soltysh, @ncdc

**Comment by liggitt** [comment]
this is post 1.1, right?

**Comment by deads2k** [comment]
Can you summarize the additional benefit we'd get by bumping levels?

**Comment by miminar** [comment]
@deads2k most notably swift support. There are customers asking for this.

@pweil- shall we push this before 1.1?

**Comment by kargakis** [comment]
> @pweil- shall we push this before 1.1?

Hm, still a big change for the registry...

**Comment by soltysh** [comment]
What about removing `code.google.com/p/go-uuid/uuid` that @kargakis brought up in #4339. I see docker is using `github.com/docker/distribution/uuid` we k8s uses `github.com/pborman/uuid`. Without removing we have 3 packages in our godeps doing the same/similar thing. We could at least get rid of the first one.

**Comment by kargakis** [comment]
> What about removing code.google.com/p/go-uuid/uuid that @kargakis brought up in #4339. I see docker is using github.com/docker/distribution/uuid we k8s uses github.com/pborman/uuid. Without removing we have 3 packages in our godeps doing the same/similar thing. We could at least get rid of the first one.

We have a couple of places in Origin where we use the code.google package, we could get replace those for start. 
```
pkg/cmd/server/origin/auth.go
13:	"code.google.com/p/go-uuid/uuid"

pkg/cmd/server/origin/auth_config.go
8:	"code.google.com/p/go-uuid/uuid"

pkg/auth/server/csrf/cookie.go
6:	"code.google.com/p/go-uuid/uuid"

pkg/auth/server/csrf/session.go
8:	"code.google.com/p/go-uuid/uuid"

pkg/generate/app/app.go
13:	"code.google.com/p/go-uuid/uuid"
```

**Comment by deads2k** [comment]
@smarterclayton This looks like a really big change to bring in now.  Is the swift support worth it?

**Comment by soltysh** [comment]
Additionally, is it possible to upgrade the docker/distribution without the biggest issue (#3333) addressed?

**Comment by pweil-** [comment]
> @pweil- shall we push this before 1.1?

I'd be worried about pulling it in at this stage in the release unless it is 100% required.  I'd prefer we hang on to it for a 3.1.1 or 3.1.2 but I'll defer to @smarterclayton

**Comment by miminar** [comment]
> What about removing code.google.com/p/go-uuid/uuid that @kargakis brought up in #4339.
Will do.
> Additionally, is it possible to upgrade the docker/distribution without the biggest issue (#3333) addressed?
Current pruning code will only be using a bit different api calls exported with still unmerged [PR#1050](https://github.com/docker/distribution/pull/1050).

**Comment by ncdc** [comment]
I don't think this should be merged until post 3.1. I also don't think this should be merged until upstream 1050 is merged. We don't want to risk deviating from what ultimately goes in upstream. Also, I'm seeing a proposed restructuring of some of the top level interfaces (https://github.com/docker/distribution/pull/1105). It might be worth getting an ETA on when those will land so we don't have to change things twice (once now, once after that lands).

**Comment by miminar** [comment]
Yeah, that upstream restructuring PR would really breaks our API pruning PR into pieces. It'd be worth to get it upstream first.

**Comment by ncdc** [comment]
@miminar can you find out the status of 1105? Do they have an ETA?

**Comment by smarterclayton** [comment]
If we don't do this for 3.1 it is *must have* (as in, top priority) for
3.1.1.

**Comment by miminar** [comment]
Their ETA is 1.10, which is not yet in their roadmap but based on their 2 months release cycle it could be the end of December. And luckily, according to today's irc discussion, they I'd like to get 1050 in before 1105.

**Comment by miminar** [comment]
Rebased to `v2.2.0`. I'll update this once more with two days old `v2.2.1` - a [bugfix release](https://github.com/docker/distribution/issues?q=milestone%3ARegistry%2F2.2.1+is%3Aclosed).

Reenabled liveness probe (issue #5900). I'd like to add registry's readiness probe once this is in.

**Comment by soltysh** [comment]
I'd appreciate if the two commits regarding uuid package changes has those package names in it.

**Comment by soltysh** [comment]
That last commit (regarding changes in OpenShift) is a bit lengthy, just by looking into the description. Could you split it into logical parts from the description?

**Comment by soltysh** [comment]
[test]

**Comment by soltysh** [comment]
The error from tests proves my argument about not updated `test/end-to-end/core.sh`:
```
[INFO] Verifying the docker-registry is up at 172.30.169.199:5000
ERROR: gave up waiting for http://172.30.169.199:5000/healthz
!!! Error in hack/../test/end-to-end/core.sh:313
```

**Comment by miminar** [comment]
Rebased to `2.2.1`.

> I'd appreciate if the two commits regarding uuid package changes has those package names in it. 

fixed.

> The error from tests proves my argument about not updated test/end-to-end/core.sh

yes, updated. Thanks for pointing that out.

**Comment by miminar** [comment]
> That last commit (regarding changes in OpenShift) is a bit lengthy,

broken up into smaller pieces

> I can agree that warning is not something you want to see at top of the log. Still global vars are 
bigger evil, imho! If something, I'd go with your 2nd option, rename our vars.

working on it. re[test]

**Comment by miminar** [comment]
Also ping @kargakis

**Comment by miminar** [comment]
Flake #6270 in build 7748.

**Comment by miminar** [comment]
I have no idea, how to name my commits to pass the `continuous-integration/travis-ci/pr` after the [749428c4a](https://github.com/openshift/origin/commit/749428c4a651faf249121e81e8cc21ff8d7826c5):

    The following UPSTREAM commits have invalid summaries:

      [f67aa54] UPSTREAM: Removed obsoleted godep code.google.com/p/go-uuid/uuid

      [ab9e95c] UPSTREAM: bump(github.com/stevvooe/resumable): 51ad44105773cafcbe91927f70ac68e1bf78f8b4

      [65a1e7d] UPSTREAM: bump(github.com/docker/distribution): e6c60e79c570f97ef36f280fcebed497682a5f37

    Examples of valid summaries:

      UPSTREAM: 12345: An origin change

      UPSTREAM: docker/docker: 12345: A docker change

      UPSTREAM: <carry>: A carried kube change

      UPSTREAM: <drop>: A dropped kube change

      UPSTREAM: revert: abcd123: docker/docker: 12345: A docker change

If I remove the `UPSTREAM: ` prefix (so I get their original title), I get:

    The following commits contain Godeps changes but aren't declared as UPSTREAM::

        [9405adb] Removed obsoleted godep code.google.com/p/go-uuid/uuid
        ...
        [28b5b42] bump(github.com/stevvooe/resumable): 51ad44105773cafcbe91927f70ac68e1bf78f8b4
        ...
        [f8eff9f] bump(github.com/docker/distribution): e6c60e79c570f97ef36f280fcebed497682a5f37

@ironcladlou any idea?

**Comment by deads2k** [comment]
> @ironcladlou any idea?

Umm... Well, we may have forgotten to allow bump commits.  @ironcladlou  can you special case that just allow any bump through until we get it sorted out more correctly?

**Comment by ironcladlou** [comment]
Yeah, let me put in a special case for bump to get @miminar moving again. 

@miminar other than the bump commit, are you clear from the examples on the format of the non-bump commit?

**Comment by miminar** [comment]
@ironcladlou still I'm not sure, what title to use for the removed dependency. Revert doesn't seem to be the case, because several commits bumped that particular package.

**Comment by ironcladlou** [comment]
@miminar #6273 will let your bump commits through and will make your shortened repo references (e.g. docker/distribution) work. @deads2k, thoughts on the summary to use for a case like f67aa54d257d95e404adfda6f6da3b5642a48837?

**Comment by pweil-** [comment]
@miminar - just a couple of questions from my review.

@soltysh @ncdc - I'd definitely like someone who has been in the registry guts a bit more than I have to take a look too.

**Comment by ncdc** [comment]
I most likely won't be able to review until Monday at the earliest.

**Comment by soltysh** [comment]
@pweil- I've already took a pass. If my comments were addressed., and I
think they will, I'm ok to merge this. I haven't seen any big issues here,
mostly nits.

**Comment by soltysh** [comment]
Now looking at travis error I got enlightened. Your bump commits should not be prefixed with UPSTREAM, just `bump(pkg): id`. UPSTREAM commits are reserved for our patches/changes, which is what you're doing in other 4 commits. The same applies to removing `code.google.com/p/go-uuid/uuid` package, no UPSTREAM prefix for this fellow.

**Comment by miminar** [comment]
@pweil- thanks for the comments. They're addressed now with [one exception](https://github.com/miminar/origin/commit/de957cf2e5a8f55147ad901d8c4b17ed270f49e5#commitcomment-14934541) which needs a bit more testing.

@soltysh:
> Your bump commits should not be prefixed with UPSTREAM, just bump(pkg): id.

Yeah, I know, but new commit-checking scripts made me [waver](https://github.com/openshift/origin/pull/5404#issuecomment-164030688) :smiley:.

**Comment by smarterclayton** [comment]
Started to do image import with auth on top of this.

**Comment by soltysh** [comment]
LGTM

**Comment by miminar** [comment]
Removed two obsolete admin routes. Fixed a re-push issue after a tag history list gets emptied.

**Comment by smarterclayton** [comment]
What is this waiting on?

On Dec 14, 2015, at 7:21 AM, OpenShift Bot <notifications@github.com> wrote:

continuous-integration/openshift-jenkins/test SUCCESS (
https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/7780/)

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/5404#issuecomment-164423852>.

**Comment by miminar** [comment]
For a magical *merge* word. I don't have merge rights.

**Comment by soltysh** [comment]
I thought we're waiting for final sing-off from @ncdc, if not I can merge it right away.

**Comment by ncdc** [comment]
Sorry, this slipped by me yesterday. It's been a bit of a crazy morning at home so far, so once I've finished eating, I'll take a look.

**Comment by miminar** [comment]
Flake #6339. re-[test]

**Comment by miminar** [comment]
I believe all the issues raised have been addressed. Is there anything else?

**Comment by ncdc** [comment]
I got pulled away and didn't thoroughly review all the commits. If you
already have an ok, don't let me hold this up.

**Comment by pweil-** [comment]
[merge]

**Comment by soltysh** [comment]
I'm gonna regret it, but yay! ;)

**Comment by kargakis** [comment]
flake https://github.com/openshift/origin/issues/6065 

[merge]

**Comment by soltysh** [comment]
I think it's a flake from #6259.

[merge]

**Comment by miminar** [comment]
The same "*flake*" #6259 for the second time. I'll try to rebase and debug locally.
